libraries/hotstuff/include/eosio/hotstuff/qc_chain.hpp

```diff
@@ -77,71 +77,261 @@ namespace eosio::hotstuff {
+         _weak_votes.reset(num_finalizers);
+         _num_finalizers = num_finalizers;
+         _state = state_t::unrestricted;
       }
 
-      void reset(const fc::sha256& proposal, size_t finalizer_size) {
-         proposal_id = proposal;
-         active_finalizers = hs_bitset{finalizer_size};
-         active_agg_sig = fc::crypto::blslib::bls_signature();
-         quorum_met = false;
+      bool add_strong_vote(const std::vector<uint8_t>& proposal_digest,
+                           size_t index,
+                           const bls_public_key& pubkey,
+                           const bls_signature& sig) {
+         assert(index < _num_finalizers);
+         if (!_strong_votes.add_vote(proposal_digest, index, pubkey, sig))
+            return false;
+         size_t weak   = num_weak();
+         size_t strong = num_strong();
+         
+         switch(_state) {
+         case state_t::unrestricted:
+         case state_t::restricted:
+            if (strong >= _quorum)
+                _state = state_t::strong;
+            else if (weak + strong >= _quorum)
+               _state = state_t::weak_achieved;
+            break;
+            
+         case state_t::weak_achieved:
+            if (strong >= _quorum)
+               _state = state_t::strong;
+            break;
+            
+         case state_t::weak_final:
+         case state_t::strong:
+            // getting another strong vote...nothing to do
+            break;
+         }
+         return true;
       }
 
-      const hs_bitset& get_active_finalizers() const {
-         assert(!active_finalizers.empty());
-         return active_finalizers;
+      bool add_weak_vote(const std::vector<uint8_t>& proposal_digest,
+                         size_t index,
+                         const bls_public_key& pubkey,
+                         const bls_signature& sig) {
+         assert(index < _num_finalizers);
+         if (!_weak_votes.add_vote(proposal_digest, index, pubkey, sig))
+            return false;
+         size_t weak   = num_weak();
+         size_t strong = num_strong();
+         
+         switch(_state) {
+         case state_t::unrestricted:
+         case state_t::restricted:
+            if (weak + strong >= _quorum)
+               _state = state_t::weak_achieved;
+            
+            if (weak >= (_num_finalizers - _quorum)) {
+               if (_state == state_t::weak_achieved)
+                  _state = state_t::weak_final;
+               else if (_state == state_t::unrestricted)
+                  _state = state_t::restricted;
+            }
+            break;
+            
+         case state_t::weak_achieved:
+            if (weak >= (_num_finalizers - _quorum))
+               _state = state_t::weak_final;
+            break;
+            
+         case state_t::weak_final:
+         case state_t::strong:
+            // getting another weak vote... nothing to do
+            break;
+         }
+         return true;
       }
```

Comment:
Please add unittests for `pending_quorum_certificate` state transitions via votes.

Reply:
Will do in a separate PR.

Reply:
The test was ready, so I just added it to this PR. Caugh one small issue too :-).